cache/cache.go

```diff
@@ -105,50 +89,50 @@ func (c *Cache) InsertSignature(s string, sig *dns.RRSIG) {
 		return
 	}
 	c.Lock()
-	defer c.Unlock()
+
 	if _, ok := c.m[s]; !ok {
 		m := ((int64(sig.Expiration) - time.Now().Unix()) / (1 << 31)) - 1
 		if m < 0 {
 			m = 0
 		}
 		t := time.Unix(int64(sig.Expiration)-(m*(1<<31)), 0).UTC()
-		e := c.l.PushFront(&elem{s, t, &dns.Msg{Answer: []dns.RR{sig}}})
-		c.m[s] = e
+		c.m[s] = &elem{t, &dns.Msg{Answer: []dns.RR{dns.Copy(sig)}}}
 	}
-	c.size += 1
-	c.shrink()
+	c.EvictRandom()
+	c.Unlock()
 }
 
-// Search returns .... and a boolean indicating if we found something
+// Search returns a dns.Msg, the expiration time and a boolean indicating if we found something
 // in the cache.
 func (c *Cache) Search(s string) (*dns.Msg, time.Time, bool) {
 	if c.capacity == 0 {
```

Comment:
This doesnt seem right. If the capacity is empty it means that we should definitely search